### PR TITLE
[HF] Model Definition Conversion Support for FLUX

### PR DESCRIPTION
**Pull request by wesleytruong**
This PR adds the `FluxStateDictAdapter`, allowing us to convert checkpoints to and from HF.

Additional changes:
- Modifies `download_hf_assets` script to support downloading diffusion-type safetensor files
- Registers Flux's `TrainSpec` in `convert_from_hf` and `convert_to_hf` so that conversion script can be reused
    - e.g. `python ./scripts/checkpoint_conversion/convert_from_hf.py ./assets/hf/FLUX.1-dev/transformer ./outputs/temp --model_name flux --model_flavor flux-dev`

Tests:
Performing KL divergence test on the forward pass of converted weights loaded in `torchtitan` and HF weights loaded with HF `FluxTransformer2DModel`, we get:
```
Average loss for test from_hf is 7.233546986222528e-13
```

Addiitonally, we can now run inference with HF weights to verify changes made in https://github.com/pytorch/torchtitan/pull/1548

### Batched Inference on TorchTitan:
| | prompt0 | prompt1 | prompt2 |
| --- | --- | --- | --- |
| no CFG | prompt0_nocfg | prompt1_nocfg | prompt2_nocfg |
| CFG | prompt0_cfg | prompt1_cfg | prompt2_cfg |

### CODE REVIEW

File: torchtitan/protocols/state_dict_adapter.py [hunk 1]
```diff
@@ -60,16 +60,23 @@ class StateDictAdapter(BaseStateDictAdapter):
...
     def __init__(self, model_args: BaseModelArgs, hf_assets_path: str | None):
         if hf_assets_path:
-            mapping_path = os.path.join(hf_assets_path, "model.safetensors.index.json")
-            try:
-                with open(mapping_path, "r") as f:
-                    hf_safetensors_indx = json.load(f)
-            except FileNotFoundError:
+            index_files = [
+                "model.safetensors.index.json",
+                "diffusion_pytorch_model.safetensors.index.json",
```
Comment (tianyu-l) [inline torchtitan/protocols/state_dict_adapter.py]: I think we shouldn't hardcode diffusion model specific file name here. If this is special treatment, let's do it in Flux's own state_dict_adapter (maybe without inheriting this `__init__`).

File: scripts/checkpoint_conversion/convert_from_hf.py [hunk 2]
```diff
@@ -9,6 +9,7 @@
 
 import torch
 import torch.distributed.checkpoint as dcp
+import torchtitan.experiments.flux  # noqa: F401
```
Comment (tianyu-l) [inline scripts/checkpoint_conversion/convert_from_hf.py]: why do we need this but didn't need to import llama?

**Review by tianyu-l — COMMENTED** [review]
Picture looks great!

Had some comments.

File: torchtitan/experiments/flux/model/state_dict_adapter.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+                hf_key = hf_abstract_key.format(layer_num)
+                combine_values.append(hf_fqn_map[hf_key])
+
+            value = torch.cat(combine_values, dim=0)
+            state_dict[tt_fqn] = value
+
+        return state_dict
+
+
+def build_flux_state_dict_adapter(
```
Comment (wwwjn) [inline torchtitan/experiments/flux/model/state_dict_adapter.py]: Do we need this function? If we want to make field in TrainSpec like ``build_xxxx" , and not plug in the class directly, we could use this function, but I didn't see it used anywhere now

File: torchtitan/experiments/flux/dataset/flux_dataset.py [hunk 4]
```diff
@@ -387,6 +387,9 @@ def __init__(
             val_timesteps = [1 / 8 * (i + 0.5) for i in range(8)]
             self.timestep_cycle = itertools.cycle(val_timesteps)
 
+        # Disable classifier free guidance for validation
+        self.job_config.training.classifier_free_guidance_prob = 0.0
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/dataset/flux_dataset.py]: I see only appearance of this field at https://github.com/pytorch/torchtitan/blob/f9e8897a6f09fbe932e4b1a4df9fc23ca08c2b5e/torchtitan/experiments/flux/dataset/flux_dataset.py#L291

You already called `super().__init__()` above, will this still take effect?

Also it sounds to me that you might have to pass this in the constructor, unless you have other better ideas?

File: torchtitan/experiments/flux/model/state_dict_adapter.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+import logging
+import os
+import re
+
+from collections import defaultdict
+from typing import Any
+
+import torch
+
+logger = logging.getLogger()
```
Comment (fegin) [inline torchtitan/experiments/flux/model/state_dict_adapter.py]: nit, it's better to put this line after the last `import` (line 21 in this case).

File: torchtitan/experiments/flux/dataset/flux_dataset.py (hunk 4 above)
Comment (wesleytruong) [reply torchtitan/experiments/flux/dataset/flux_dataset.py]: You're right in the above implementation this line unintentionally disabled using dropout during training as well, since the job_config is globally changed.

I switched this override to happen inside flux `validator.validate` now and switch back at the end

**Review by tianyu-l — APPROVED** [review]
didn't read into conversion map details, but the organization looks good to me.
